File: java/src/sorter/TournamentPQTest.java
package sorter;

import java.util.ArrayList;
import java.util.List;

import org.junit.Assert;
import org.junit.jupiter.api.Test;

class TournamentPQTest {

    @Test
    void testFiveProviders() {
        int numProviders = 4;
        int numItemsPerList = 10;
    
        Provider p = new InOrderGenerator(numProviders * numItemsPerList, 16);
        
        @SuppressWarnings("unchecked")
        List<Record>[] lists = new List[numProviders];
        
        for(int i=0; i<numProviders ; i++) {
            lists[i] = new ArrayList<Record>(numItemsPerList);
        }
        
        for(int j=0 ; j<numItemsPerList ; j++) {
            for(int i=numProviders-1 ; i>=0 ; i--) {
                lists[i].add(p.next());
            }
        }
        
        Provider[] providers = new Provider[numProviders];
        for(int i=0; i<numProviders ; i++) {
            providers[i] = new ArrayProvider("Provider-"+i, lists[i].iterator());
        }
        
        TournamentPQ pq = new TournamentPQ(providers);
        Witness w = new Witness(pq);
        Printer pr = new Printer(w, "After Tournament");
        Consumer c = new Consumer(pr);
        c.consume();
        
        Assert.assertTrue(w.isSorted());  
    }

}

File: Testing/TournamentPqTest.cpp
#include "TournamentPqTest.h"
#include "Testing/TestProviders/InOrderGenerator.h"
#include "../TournamentPQ.h"
#include "../Consumer.h"
#include "Witness.h"
#include <iostream>
#include <algorithm>
#include <vector>
#include <cassert>
using namespace std;


void TournamentPqTest::doTest(int numProviders) {
    int numItemsPerList = 10;

    InOrderGenerator i(numProviders * numItemsPerList, 16);
    Provider p = i;
    Witness before(p);


    vector<vector<Record>> lists(numProviders);

    for(int i=0; i<numProviders ; i++) {
        vector<Record> l(numItemsPerList);
        lists[i] = l;
    }

    for(int j=0 ; j<numItemsPerList ; j++) {
        for(int i=numProviders-1 ; i>=0 ; i--) {
            Record* ptr = before.next();
            Record r = *ptr;
            lists[i].push_back(r);
        }
    }

    vector<Provider> providers(numProviders);
    for(int i=0; i<numProviders ; i++) {
        //TODO: fix iterator
        ArrayProvider p("Provider-"+i, lists[i].iterator())
        providers[i] = p;
    }


    TournamentPQ pq(providers, 8, providers.size());
    Witness w(pq);
    Consumer c(w);
    c.consume();

    assert(w.isSorted);
    assert(("The count of the before witness did have the same checksum as the w witness but should have", before.getCrc() == w.getCrc()));
}

void TournamentPqTest::testVariousNumberOfProviders() {
    for(int i=1; i<100; i++) {
        doTest(i);
    }
}
